core/src/connection/error.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::transport::TransportError;
+use std::{io, fmt};
+
+/// Errors that can occur in the context of a pending or established `Connection`.
+#[derive(Debug)]
+pub enum ConnectionError<THandlerErr, TTransErr> {
```

Comment:
Will do, thanks!

Reply:
Since the use of `thiserror` for such an error type with type parameters requires trait bounds to be present on the error type itself, which in turn propagate through the code base, there actually does not seem to be a lot of gain in using it here. Nevertheless the experiment can be seen [here](https://github.com/romanb/rust-libp2p/commit/9466a4526a62b2cdc8bb58a9ef6258c91e615761).

Reply:
Thanks for giving it a go.